[#20] Add logger rotation action

sphaso opened the pull request:
(will) Resolve #20 

I'm currently trying to solve the compilation error, I guess I should combine `putStrLn` somewhere but the type tetris is failing me. I opened the PR anyway to check if you like the approach or if you have any other remark, please do tell.

ChShersh commented on co-log/src/Colog/Actions.hs:
```diff
@@ -92,9 +102,37 @@ withLogRotation
-        withFile path AppendMode writeFileLoop
--}
+    rotationAction = liftLogIO $ LogAction $ (\msg ->
+        withFile path AppendMode (flip (unLogAction . mkAction) msg))
```
You got the idea right. But returns `rotationAction` is not valid, unfortunately. You return this action and it's supposed to do all the dirty work. So on every call to `LogAction` this function should output line to the file, check resulted length of the file and rename files if needed. Currently you're performing renaming only once, at the beginning of `withLogRotation` while in reality this should be done on every call to `rotationAction`.

The difficulty in implementing this function is checking the size of the file each time, keep outputting to `Handle` instead of reopening file each time and use new `Handle` each time the size of the file is bigger than the given `Limit`

ChShersh replied:
> Would we expect the user to have a partially applied withLogRotation and pass cont each time they want to log?

The use-case is that you call `withLogRotation` only once in your program to create properly configured `LogAction` and then you use only that action. Compare with more simple example: `withFile`. It's more efficient to open file only once and use `Handle` to write output instead of calling `appendFile` on every message. So you open file only once in your program, and then you use given action to output the result. Same with `withLogRotation`. You specify configuration parameters and then you can use single action to output messages, and this action handles file renaming and size check for you.

ChShersh commented on co-log/src/Colog/Actions.hs:
```diff
@@ -62,8 +67,13 @@ withLogTextFile path action = withFile path AppendMode $ action . logTextHandle
-{-
-data Limit = LimitTo Natural | Unlimited
+
+data Limit = LimitTo Natural | Unlimited deriving (Eq)
```
Actually, if you write constructors in this particular order, the derived `Ord` instances will be correct, no need to write it manually

ChShersh commented:
@sphaso Thank you for continuing to work on this! The issue is actually is pretty difficult, so no easy way to resolve it...

> Do you think msg can be kept polymorphic? I see other withFile examples have it constrained to either String, ByteString or Text...

I think that `msg` can be kept polymorphic. The idea is that `withLogRotation` action will be called like this:

```haskell
main :: IO ()
main = withLogRotation (Limit 3) (Limit 5) "node.log" removeFile logByteStringHandle $ \action ->
    ... the rest of the program ...
```

And instead of passing `logByteStringHandle` I can pass `logTextHandle` or `logStringHandle`. So there actually shouldn't be any logic related to the type of `msg` rather than just calling given action on the `msg`. You can write `msg` to `Handle` and get size of `Handle` after outputting message. So no need to calculate size of `msg`.

ChShersh commented:
@sphaso I have an idea on how to implement this. I think that this can be done only using `IORef Handle` and manually working with `open` and `hClose` functions carefully cleaning up resources. I don't see way to implement this using `withFile` function (though it would be really nice to do so). 

And I think that the issue should be separated into smaller ones... Like:

1. Implement `Limit` data type, instances for this data type and unit tests for `Ord` instance.
2. Implement function that walks over files with lower numbers and applies given cleaning function to them.
3. Implement function that writes to `Handle` and checks the size of the `Handle` and calls previous function if case if file exceeds limit.
4. Combine everything into `withLogRotation` function.

sphaso commented:
@ChShersh Hi, I've arrived to a similar conclusion (not being able to use `withFile`, but using `open` instead). I'll implement the steps you mentioned, although I think 2 and 3 might be merged into a single function? (that's what I did in the code).     
Right now I'm a bit stuck with the monad stack types, but I feel very close to a conclusion!

sphaso commented:
@ChShersh thanks to the help of @alpacaaa I was able to have a compiling version. I'm not sure what you mean by
> Implement Limit data type, instances for this data type and unit tests for Ord instance.

Do you mean that we should have `data Limit a = Limit a | Unlimited` instead of `Limit Natural`? Does it need to have its own class?

Regarding point 2 and 3, as you can tell from the code, I make both checks every time, since I supposed both conditions need to pass before writing to the file (both size and file numbers need to be below the limit).

ChShersh commented:
@sphaso 

>

I meant exactly what you did, so good job 👍 

```haskell
data Limit = LimitTo Natural | Unlimited deriving (Eq, Ord)
```

I've checked the derived instance and it works as expected 👍

<img width="430" alt="screenshot 2018-10-29 at 12 20 27 pm" src="https://user-images.githubusercontent.com/4276606/47637960-2f4eb780-db98-11e8-928a-9165013792d0.png">

sphaso commented:
@ChShersh work done since last commit:
- put `ScopedTypeVariables` in cabal (and removed it from two other modules in `co-log`)
- exported `Limit` and `withLogRotation` from Action module
- implemented the functions you suggested and modified the code accordingly, it looks much cleaner!
- removed the `IORef` from `withLogRotation` parameters (we had added it to the parameters because we though that by creating it into the body it would make a new one each time the `cont` would be applied)
- used `bracket` (very nice addition too!)
Now I'm battling against the usual error: `Couldn't match type ‘m’ with ‘IO’`
but I imagine you might want to change more code, so I won't focus too much on that.

As an aside, I wanted to thank you guys for making your projects so beginner-friendly, especially for the hacktoberfest. It's been a blast and I hope to continue contributing in the future. My Haskell knowledge sure improved a lot!

ChShersh commented on co-log/src/Colog/Actions.hs:
```diff
@@ -92,9 +102,64 @@ withLogRotation
+          handle <- readIORef refHandle
+          sizeLimitReached <- isFileSizeLimitReached sizeLimit handle
+
+          if sizeLimitReached then do
```
Actually, the logic can be slightly simpler. It's okay to always write `msg` to the existing `Handle` and perform renaming only after message is written. Like this:

```haskell
outputMsg msg handle
whenM (isFileSizeLimitReached sizeLimit handle) cleanupAndRotate
```

Also, you need to cleanup only when you perform logger rotation. Don't need to cleanup every time on each message.

sphaso commented on co-log/src/Colog/Actions.hs:
```diff
@@ -92,9 +102,64 @@ withLogRotation
+
+-- creates list of files with indices who are older on given Limit than the latest one
+getOldFiles :: Limit -> FilePath -> IO [FilePath]
+getOldFiles limit path = do
```
Thank you, this is very informative! I guess it was fair to assume I should have understood this from the start

sphaso commented:
@ChShersh sorry for the sloppy style. This time I used `stylish-haskell` and double checked the style guidelines, so everything should be properly aligned and ordered.

I think I implemented all the changes you mentioned. The code looks so much cleaner now :) 
I have some questions:

**1. type inference**
Why does this work
```Haskell
    cleanupAndRotate refHandle = liftIO $ do
      h <- readIORef refHandle
      hClose h
```
but `hClose <$> readIOref refHandle` gives an error for an ambiguos MonadIO?
```
Could not deduce (MonadIO m0) arising from a use of ‘hClose’
      from the context: MonadIO m
```
It took me a while to figure out that just breaking the function could work, but it was sheer intuition, not knowledge.

**2. algebra**
There's two points where I check if `Limit` is `Unlimited` or `LimitTo n`, does this warrant having a `Functor` instance of `Limit`?

ChShersh commented:
@sphaso 

> Why does this work .. but ... gives an error for an ambiguos MonadIO?

I actually don't know. Depends on the context. But `hClose <$> readIOref refHandle` is not the equivalent of

```haskell
    cleanupAndRotate refHandle = liftIO $ do
      h <- readIORef refHandle
      hClose h
```

it should be `readIORef refHandle >>= hClose`. I think because of that type inference got broken.

> 2. algebra
> There's two points where I check if Limit is Unlimited or LimitTo n, does this warrant having a Functor > instance of Limit?

Well, `Limit` data type is not polymorphic so having `Functor` instance for it won't help. If you pattern-match on `Limit` constructors multiple types, you can implement function:

```haskell
limit :: a -> (Natural -> a) -> Limit -> a
```

which is called `Limit`-eliminator (similar to `maybe :: b -> (a -> b) -> Maybe a -> b` which is called `Maybe`-eliminator).

If having such `limit` function helps you then it's okay 🙂

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,112 @@
+  where
+    rotationAction :: IORef Handle -> LogAction m msg
+    rotationAction refHandle
+      = LogAction $ \msg -> do
```
I think that this function can be written slightly shorter. No need to move on separate line:

```haskell
rotationAction refHandle = LogAction $ \msg -> do
    handle <- ...
```

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,112 @@
+        unLogAction (mkAction handle) msg
+
+        whenM
+          (liftIO $ isFileSizeLimitReached sizeLimit handle)
```
You can put `liftIO` in front of `do` in `isFileSizeLimitReached` function to make this `whenM` call look slightly nicer.

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,112 @@
+      h <- readIORef refHandle
+      hClose h
+      maxN <- maxFileIndex path
+      renameFileToNumber (succ maxN) path
```
`succ` is slower than `+ 1`

```suggestion
      renameFileToNumber (maxN + 1) path
```

sphaso replied:
This is surprising... any idea why?

ChShersh replied:
@sphaso `succ` checks for overflow. That's why is slower. You might be surprised by the following behavior as well:

```haskell
ghci> succ (maxBound :: Int)
*** Exception: Prelude.Enum.succ{Int}: tried to take `succ' of maxBound
```

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,112 @@
+      oldFiles <- getOldFiles filesLimit path
+      mapM_ cleanup oldFiles
+      newHandle <- openFile path AppendMode
+      modifyIORef' refHandle (const newHandle)
```
```suggestion
      writeIORef refHandle newHandle
```

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,112 @@
+renameFileToNumber n path = D.renameFile path (path POS.<.> show n)
+
+-- if you give it name like `node.log.4` then it returns `Just 4`
+logFileIndex :: FilePath -> Maybe Int
```
I feel that code can be simplified if instead of `Maybe Int` you will return `Maybe Natural` here

sphaso replied:
This is a brilliant intuition!

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,112 @@
+    currentMaxN <- maxFileIndex path
+    files <- D.listDirectory (POS.takeDirectory path)
+    let tuple = map (\a -> (a, toInteger <$> logFileIndex a)) files
+    pure $ map fst $ filter (maybe False (isOldFile currentMaxN) . snd) tuple
```
I feel like this two lines of code can be simplified. For example, you can use `mapMaybe` function to get rid of `maybe False` and `filter`

sphaso replied:
I would need a function like:
`sequenceC :: (Traversable t, Monad m) => t (a, m b) -> t (m (a, b))`
I could write it myself, but it would complicate the code just to clean these two lines. Any idea?

ChShersh replied:
Okay, it's usually helps to write types of functions to see how it can be simplified :slightly_smiling_face: Usually in such situations Haskell reminds this child game where you have objects of different shapes and you need to put them inside box where holes of those shapes should match :slightly_smiling_face:  

Context:

```haskell
files :: [FilePath]
logFileIndex :: FilePath -> Maybe Natural
isOldFile maxN :: Natural -> Bool
mapMaybe :: (a -> Maybe b) -> [a] -> [b]

<goal> :: [FilePath] -- by `boolean predicate
```

The idea is to implement couple functions:

```haskell
takeFileIndex  :: FilePath -> Maybe (FilePath, Natural)
guardFileIndex :: (FilePath, Natural) -> Maybe FilePath
-- should be quite easy using the existing functions
```

Then you will have:

```haskell
mapMaybe (takeFileIndex >=> guardFileIndex) files :: [FilePath]
```

No need to play with `Bool`, I personally like `mapMaybe` function, it's really powerful.

sphaso replied:
Wow this is amazing. I'm seeing a pattern in these refactorings: when things look like they can't compose, break them down :)

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,112 @@
+    isOldFile maxN n = case limit of
+                         Unlimited -> False
+                         LimitTo l -> n < toInteger maxN - toInteger l
+
```
Looks like we need `hintman` here :suspect:

sphaso commented:
@ChShersh Thanks again for the review. I think I fixed most of comments, I only left 2 where I have some doubts :)

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,112 @@
+    pure $ map fst $ filter (maybe False (isOldFile currentMaxN) . snd) tuple
+  where
+    isOldFile :: Int -> Integer -> Bool
+    isOldFile maxN n = case limit of
```
@sphaso Makes sense. I'm okay with having separate function for that :+1: For me it's hard to perform such kind of math in my head. Thank you for writing the equations!

sphaso replied:
not a problem, I had to write it down myself :)

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,118 @@
+-- This functionality is not to be considered stable
+-- or ready for production use. While we enourage you
+-- to try it out and report bugs, we cannot assure you
+-- that everything will work as advertised :)
```
Very nice text!

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,118 @@
+isLimitedBy _ Unlimited = False
+isLimitedBy size (LimitTo limit)
+  | size <= 0 = False
+  | otherwise = (toInteger limit :: Integer) > size
```
Should work without explicit type annotation, type inference works :+1: 
```suggestion
  | otherwise = toInteger limit > size
```

ChShersh commented on co-log/src/Colog/Rotation.hs:
```diff
@@ -0,0 +1,118 @@
+      newHandle <- openFile path AppendMode
+      writeIORef refHandle newHandle
+
+isLimitedBy :: Integer -> Limit -> Bool
```
I think, it's a good idea to specify explicitly in the comment for this function whether `isLimitedBy n limit` means `n < limit` or `n <= limit`.

sphaso commented:
@ChShersh more refactoring has been done :)
